h2/src/main/org/h2/command/Parser.java

```diff
@@ -5021,50 +5179,66 @@ private TableView parseSingleCommonTableExpression() {
-                // use the passed in column name if supplied, otherwise use alias (if used) otherwise use column name
-                // derived from column expression
-                String columnName;
-                if (cols != null){
-                    columnName = cols[i];
-                } else if (columnExp.getAlias()!=null){
-                    columnName = columnExp.getAlias();
-                }
-                else{
-                     columnName =  columnExp.getColumnName();
-                }
-                columnTemplateList.add(new Column(columnName,
-                        columnExp.getType()));
-            }
+            columnTemplateList = createQueryColumnTemplateList(cols, withQuery, querySQLOutput);
+
         } finally {
             session.removeLocalTempTable(recursiveTable);
         }
+        TableView view = createTemporarySessionView(tempViewName, querySQLOutput[0], columnTemplateList,true/*allowRecursiveQueryDetection*/, true);
+        return view;
+    }
+
+    private List<Column> createQueryColumnTemplateList(String[] cols, Query withQuery, String[] querySQLOutput) {
+        List<Column> columnTemplateList = new ArrayList<Column>();
+        withQuery.prepare();
+        querySQLOutput[0] = StringUtils.cache(withQuery.getPlanSQL());
+        ArrayList<Expression> withExpressions = withQuery.getExpressions();
+        for (int i = 0; i < withExpressions.size(); ++i) {
+            Expression columnExp = withExpressions.get(i);
+            // use the passed in column name if supplied, otherwise use alias (if used) otherwise use column name
+            // derived from column expression
+            String columnName;
+            if (cols != null){
+                columnName = cols[i];
+            } else if (columnExp.getAlias()!=null){
+                columnName = columnExp.getAlias();
+            }
+            else{
+                 columnName =  columnExp.getColumnName();
+            }
+            columnTemplateList.add(new Column(columnName,
+                    columnExp.getType()));
+        }
+        return columnTemplateList;
+    }
+
+    private TableView createTemporarySessionView(String tempViewName,  String querySQL,
+            List<Column> columnTemplateList, boolean allowRecursiveQueryDetection, boolean addViewToSession) {
+        Schema schema = getSchemaWithDefault();
         int id = database.allocateObjectId();
         // No easy way to determine if this is a recursive query up front, so we just compile
         // it twice - once without the flag set, and if we didn't see a recursive term,
         // then we just compile it again.
         TableView view = new TableView(schema, id, tempViewName, querySQL,
                 parameters, columnTemplateList.toArray(new Column[0]), session,
-                true/* recursive */, false);
-        if (!view.isRecursiveQueryDetected()) {
+                allowRecursiveQueryDetection/* recursive */, false);
```

Comment:
can remove the 
   /* recursive */
commnent now, I put that in for boolean params when passing a literal, to make it more obvious what the param is for